### PR TITLE
[Broker] Support disabling non-TLS service ports

### LINKED ISSUE #11548: Pulsar server non-tls ports are still listening when TLS in-transit encryption is enabled
**Describe the bug**
I'm testing out transport encryption using TLS by following the document instruction here:
https://pulsar.apache.org/docs/en/2.7.2/security-tls-transport/

I'm expecting that when TLS is enabled, when broker is starting up, only TLS ports (8443 and 6551) are listening. But it looks like non-TLS ports (8080 and 6650) are also enabled. See the command line output below:

```
$ sudo netstat -ntlp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:2181            0.0.0.0:*               LISTEN      16109/java
tcp        0      0 0.0.0.0:9990            0.0.0.0:*               LISTEN      16109/java
tcp        0      0 0.0.0.0:8010            0.0.0.0:*               LISTEN      16109/java
tcp        0      0 0.0.0.0:3181            0.0.0.0:*               LISTEN      16676/java
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      17362/java
tcp        0      0 127.0.0.53:53           0.0.0.0:*               LISTEN      877/systemd-resolve
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      9105/sshd
tcp        0      0 0.0.0.0:6650            0.0.0.0:*               LISTEN      17362/java
tcp        0      0 0.0.0.0:8443            0.0.0.0:*               LISTEN      17362/java
tcp        0      0 0.0.0.0:6651            0.0.0.0:*               LISTEN      17362/java
tcp        0      0 0.0.0.0:8000            0.0.0.0:*               LISTEN      16676/java
```

The only broker configuration changes are those TLS related, as below.

```
$ diff conf/broker.conf ~/pulsar_config/broker.conf.orig
32c32
< brokerServicePortTls=6651
---
> brokerServicePortTls=
38c38
< webServicePortTls=8443
---
> webServicePortTls=
472c472
< tlsEnabled=true
---
> tlsEnabled=false
478c478
< tlsCertificateFilePath=/opt/pulsar/security/tls/broker.cert.pem
---
> tlsCertificateFilePath=
481c481
< tlsKeyFilePath=/opt/pulsar/security/tls/broker.key-pk8.pem
---
> tlsKeyFilePath=
487c487
< tlsTrustCertsFilePath=/opt/pulsar/security/tls/ca.cert.pem
---
> tlsTrustCertsFilePath=
```

**To Reproduce**
1. Download Pulsar 2.7.2 release tarball and extract it to a folder, e.g. /opt/pulsar
2. Create TLS certificates ( by following Pulsar doc: https://pulsar.apache.org/docs/en/2.7.2/security-tls-transport/#broker-configuration
> put the root certificate (ca.cert.pem), the broker key (broker.key-pk8.pem), and the broker certificate (broker.cert.pem) into a sub-folder of Pulsar install: /opt/pulsar/security/tls
3. Configure broker.conf, as below. 
```
brokerServicePort=6650
brokerServicePortTls=6651
webServicePort=8080
webServicePortTls=8443
tlsEnabled=true
tlsCertificateFilePath=/opt/pulsar/security/tls/broker.cert.pem
tlsKeyFilePath=/opt/pulsar/security/tls/broker.key-pk8.pem
tlsTrustCertsFilePath=/opt/pulsar/security/tls/ca.cert.pem
```
4. Restart Pulsar server and we'll see that Pulsar server is listening on all 4 ports (8080, 8443, 6650, and 6651).

I also tried explicitly removing non-TLS ports in the configuration file and only keep TLS ports, as below. 
```
brokerServicePort=
brokerServicePortTls=6651
webServicePort=
webServicePortTls=8443
tlsEnabled=true
tlsCertificateFilePath=/opt/pulsar/security/tls/broker.cert.pem
tlsKeyFilePath=/opt/pulsar/security/tls/broker.key-pk8.pem
tlsTrustCertsFilePath=/opt/pulsar/security/tls/ca.cert.pem
```

If I do so, I can't start broker. The error in **pulsar-broker-xxx.out** is as below:
```
[AppClassLoader@18b4aac2] info AspectJ Weaver Version 1.9.2 built on Wednesday Oct 24, 2018 at 15:43:33 GMT
[AppClassLoader@18b4aac2] info register classloader sun.misc.Launcher$AppClassLoader@18b4aac2
[AppClassLoader@18b4aac2] info using configuration file:/opt/pulsar/lib/org.apache.pulsar-pulsar-zookeeper-utils-2.7.2.jar!/META-INF/aop.xml
[AppClassLoader@18b4aac2] info using configuration file:/opt/pulsar/lib/org.apache.pulsar-pulsar-zookeeper-2.7.2.jar!/META-INF/aop.xml
[AppClassLoader@18b4aac2] info register aspect org.apache.pulsar.broker.zookeeper.aspectj.ClientCnxnAspect
[AppClassLoader@18b4aac2] info register aspect org.apache.pulsar.zookeeper.FinalRequestProcessorAspect
[AppClassLoader@18b4aac2] info register aspect org.apache.pulsar.zookeeper.ZooKeeperServerAspect
2021-08-03 08:56:33,926 [sun.misc.Launcher$AppClassLoader@18b4aac2] error Uncaught exception in thread main: No value present
```

**Expected behavior**
Only TLS ports (8443 and 6551) are listening. 

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Java: 1.8.0_292

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by lhotari**
Fixes #11548

### Motivation

Disabling non-TLS service ports isn't possible at the moment. Issue #11548 has been reported and contains instructions to reproduce the issue.

### Modifications

- Make changes to supports disabling non-TLS service ports in `broker.conf`/`standalone.conf` by providing empty values:  
```
brokerServicePort=
webServicePort=
``` 
  - there were a few locations where the Optional port value was not checked for existence.
  - embedded Function Worker Service didn't work with TLS only broker. Fix issues in configuring Function Worker Service.
- add support for MockedPulsarServiceBaseTest tests to disable non-TLS ports.

### CODE REVIEW

**Review by sijie — COMMENTED** [review]
I am not sure how does your PR solve #11548. From what I can see, you are basically changing the function worker service and tests. There is nothing related to brokers. Can you clarify how does your PR fix #11548?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java [hunk 1]
```diff
@@ -1122,7 +1122,7 @@ private String getBrokerAddress() {
         return String.format("%s:%s", pulsar.getAdvertisedAddress(),
                 pulsar.getConfiguration().getWebServicePort().isPresent()
                         ? pulsar.getConfiguration().getWebServicePort().get()
-                        : pulsar.getConfiguration().getWebServicePortTls());
+                        : pulsar.getConfiguration().getWebServicePortTls().get());
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:1125]: If the Optional has no value `get` will throw a RuntimeException.
What about using orElse(null).
This case won't probably happen but it is better to not risk

File: pulsar-websocket/src/main/java/org/apache/pulsar/websocket/service/ProxyServer.java [hunk 2]
```diff
@@ -120,7 +122,9 @@ public void addRestResources(String basePath, String javaPackages, String attrib
     }
 
     public void start() throws PulsarServerException {
-        log.info("Starting web socket proxy at port {}", conf.getWebServicePort().get());
+        log.info("Starting web socket proxy at port {}", String.join(",", Arrays.stream(server.getConnectors())
+                .map(ServerConnector.class::cast).map(ServerConnector::getPort).map(Object::toString)
+                .collect(Collectors.toList())));
```
Comment (eolivelli) [inline pulsar-websocket/src/main/java/org/apache/pulsar/websocket/service/ProxyServer.java:127]: Nit: there is a Collectors.joining utility method

**Review by eolivelli — APPROVED** [review]
Overall looks good to me.
I left one comment PTAL

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java (hunk 1 above)
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:1125]: The same pattern is used elsewhere. There is a check in Broker startup that if webServicePort is Optional.empty, then webServicePortTls must be set. It should be fine to have the code as is.

**Comment by lhotari** [comment]
> I am not sure how does your PR solve #11548. From what I can see, you are basically changing the function worker service and tests. There is nothing related to brokers. Can you clarify how does your PR fix #11548?

@sijie This is the first issue to hit with "pulsar standalone":
```
12:11:04.314 [main] ERROR org.apache.pulsar.broker.PulsarService - Failed to start Pulsar service: No value present
java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:148) ~[?:?]
	at org.apache.pulsar.broker.loadbalance.NoopLoadManager.start(NoopLoadManager.java:55) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.startLoadManagementService(PulsarService.java:1047) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:751) [pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandalone.start(PulsarStandalone.java:296) [pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandaloneStarter.main(PulsarStandaloneStarter.java:131) [pulsar-broker.jar:2.9.0-SNAPSHOT]
12:11:04.314 [main] ERROR org.apache.pulsar.PulsarStandaloneStarter - Failed to start pulsar service.
org.apache.pulsar.broker.PulsarServerException: java.util.NoSuchElementException: No value present
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:796) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandalone.start(PulsarStandalone.java:296) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandaloneStarter.main(PulsarStandaloneStarter.java:131) [pulsar-broker.jar:2.9.0-SNAPSHOT]
Caused by: java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:148) ~[?:?]
	at org.apache.pulsar.broker.loadbalance.NoopLoadManager.start(NoopLoadManager.java:55) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.startLoadManagementService(PulsarService.java:1047) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:751) ~[pulsar-broker.jar:2.9.0-SNAPSHOT]
	... 2 more

```

Perhaps the PR description is currently confusing. In PR #11548, the problem is that when you set non-tls ports to an empty value, exceptions will happen. I tested with Pulsar development version by building Pulsar locally and appending these settings to `standalone.conf`:
```
brokerServicePort=
brokerServicePortTls=6651
webServicePort=
webServicePortTls=8443
brokerClientTlsEnabled=true
tlsEnabled=true
tlsAllowInsecureConnection=true
tlsCertificateFilePath=pulsar-broker/src/test/resources/authentication/tls-http/broker.cert.pem
tlsKeyFilePath=pulsar-broker/src/test/resources/authentication/tls-http/broker.key-pk8.pem
tlsTrustCertsFilePath=pulsar-broker/src/test/resources/authentication/tls-http/ca.cert.pem
``
and then starting Pulsar with `./bin/pulsar standalone`. I simply kept iterating until the problems were fixed. 

I also added unit tests to verify that an empty port value gets converted to Optional.empty. The unit test for TLS without non-TLS ports enabled is also added to verify that in unit tests. That's the reason to add these tests. I added these tests before starting to work on testing it manually with Pulsar standalone.

The reason why the change to Function Worker service is required is that "useTls" setting is deprecated in configuration:
https://github.com/apache/pulsar/blob/047fb6a3ca0cf9c137a5597339e20247d4c61cdf/pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/worker/WorkerConfig.java#L357-L360
The same applies to "tlsEnabled" in broker.conf. [It has been marked as deprecated](https://github.com/apache/pulsar/blob/ddeae659be1f80f428fdb8c82c7a9a0c931c81fe/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java#L1020-L1021).
tlsEnabled was deprecated by PR #2988 many years ago.
At first I was testing without setting tlsEnabled=true and that's why I hit the issue and needed to make the changes.

@sijie I didn't record all of the exceptions that happen without the changes and what this PR fixes. Is that necessary for accepting this PR?

**Comment by lhotari** [comment]
@sijie This was the 2nd exception that this PR fixed in Pulsar Standalone (with TLS ports only config mentioned in my previous comment):

```
12:29:46.424 [main] ERROR org.apache.pulsar.functions.worker.PulsarWorkerService - Error Starting up in worker
java.lang.NullPointerException: null
	at org.apache.pulsar.client.admin.internal.PulsarAdminImpl.<init>(PulsarAdminImpl.java:189) ~[pulsar-client-admin-original.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.client.admin.internal.PulsarAdminBuilderImpl.build(PulsarAdminBuilderImpl.java:47) ~[pulsar-client-admin-original.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.functions.worker.WorkerUtils.getPulsarAdminClient(WorkerUtils.java:221) ~[pulsar-functions-worker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.functions.worker.WorkerUtils.getPulsarAdminClient(WorkerUtils.java:196) ~[pulsar-functions-worker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.functions.worker.PulsarWorkerService$1.newPulsarAdmin(PulsarWorkerService.java:144) ~[pulsar-functions-worker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.functions.worker.PulsarWorkerService.start(PulsarWorkerService.java:433) [pulsar-functions-worker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.startWorkerService(PulsarService.java:1575) [pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.broker.PulsarService.start(PulsarService.java:764) [pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandalone.start(PulsarStandalone.java:296) [pulsar-broker.jar:2.9.0-SNAPSHOT]
	at org.apache.pulsar.PulsarStandaloneStarter.main(PulsarStandaloneStarter.java:131) [pulsar-broker.jar:2.9.0-SNAPSHOT]
```

**Comment by lhotari** [comment]
without making any changes, disabling non-TLS service ports works for Pulsar Stanadlone when Function Worker is disabled and when `loadManagerClassName=org.apache.pulsar.broker.loadbalance.impl.ModularLoadManagerImpl` setting is used.  
I verified by appending these lines to conf/standalone.conf
```
brokerServicePort=
brokerServicePortTls=6651
webServicePort=
webServicePortTls=8443
brokerClientTlsEnabled=true
tlsEnabled=true
tlsAllowInsecureConnection=true
tlsCertificateFilePath=pulsar-broker/src/test/resources/authentication/tls-http/broker.cert.pem
tlsKeyFilePath=pulsar-broker/src/test/resources/authentication/tls-http/broker.key-pk8.pem
tlsTrustCertsFilePath=pulsar-broker/src/test/resources/authentication/tls-http/ca.cert.pem
loadManagerClassName=org.apache.pulsar.broker.loadbalance.impl.ModularLoadManagerImpl
```
and starting with `./bin/pulsar standalone --no-functions-worker --no-stream-storage`

@sijie This is the reason why this fix contains some changes to Functions worker configuration.

**Review by merlimat — APPROVED** [review]

**Comment by Anonymitaet** [comment]
Thanks for your contribution. Do we need to update docs here https://pulsar.apache.org/docs/en/next/reference-configuration/#broker?

**Review by michaeljmarshall — APPROVED** [review]
LGTM

It's worth adding that the helm chart already creates the pulsar broker service in such a way that either the non-TLS port or the TLS port (exclusively) is open for the `pulsar` protocol: https://github.com/apache/pulsar-helm-chart/blob/c3e4ea272b15d7e806d2585059686e8270b98b67/charts/pulsar/templates/broker-service.yaml#L32-L45

**Comment by lhotari** [comment]
> Thanks for your contribution. Do we need to update docs here https://pulsar.apache.org/docs/en/next/reference-configuration/#broker?

@Anonymitaet Yes, it would be useful to add docs that it's possible to disable the non-TLS ports `brokerServicePort` and `webServicePort` by providing an empty value. In this case it's necessary to specify `brokerClientTlsEnabled=true` and either set `tlsAllowInsecureConnection=true` or configure `brokerClientTlsEnabledWithKeyStore=true` and the related settings `brokerClientTlsTrustStore` and `brokerClientTlsTrustStorePassword`. I'll add this to the docs.

**Comment by lhotari** [comment]
@Anonymitaet I added docs to `security-tls-keystore.md` file. Please review

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java (hunk 1 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/SimpleLoadManagerImpl.java:1125]: makes sense

File: site2/docs/security-tls-keystore.md [hunk 3]
```diff
@@ -131,6 +131,13 @@ brokerClientTlsTrustStorePassword=clientpw
 
 NOTE: it is important to restrict access to the store files via filesystem permissions.
 
+When TLS is configured on the broker, the non-TLS ports `brokerServicePort` and `webServicePort` can be disabled by providing an empty value in configuration.
+```
+brokerServicePort=
+webServicePort=
+```
+In this case it's mandatory to specify `brokerClientTlsEnabled=true`, `brokerClientTlsEnabledWithKeyStore=true` and the related configuration properties `brokerClientTlsTrustStore` and `brokerClientTlsTrustStorePassword`.
```
Comment (Anonymitaet) [inline site2/docs/security-tls-keystore.md:139]:
```suggestion
In this case, you need to set the following configurations.

```conf
brokerClientTlsEnabled=true // Set this to true
brokerClientTlsEnabledWithKeyStore=true  // Set this to true
brokerClientTlsTrustStore= // Set this to your desired value
brokerClientTlsTrustStorePassword= // Set this to your desired value
```
```

**Review by Anonymitaet — COMMENTED** [review]
Thanks for your contribution. I've made some comments, PTAL.

**Comment by lhotari** [comment]
> Thanks for your contribution. I've made some comments, PTAL.

@Anonymitaet  thanks for the suggestions. I committed them. Please PTAL

**Review by Anonymitaet — APPROVED** [review]

**Comment by lhotari** [comment]
PR for cherry-picking to branch-2.7 is #11724

**Comment by lhotari** [comment]
#20535 is also required for supporting disabling of non-TLS service ports.

**Pull request merged**